store/postgres/src/chain_store.rs

```diff
@@ -1813,6 +1841,120 @@ impl ChainStoreTrait for ChainStore {
+            RecentBlocksCache {
+                inner: RwLock::new(Inner {
+                    blocks: HashMap::new(),
+                    capacity,
+                }),
+            }
+        }
+
+        pub fn set_chain_head(&self, chain_head: BlockPtr) {
+            // We invalidate the entire cache when the chain head changes.
+            let mut inner = self.inner.write();
+            inner.blocks.clear();
+            inner.blocks.insert(0, (chain_head, None));
+        }
+
+        pub fn chain_head_ptr(&self) -> Option<BlockPtr> {
+            self.inner.read().blocks.get(&0).map(|p| p.0.clone())
+        }
+
+        pub fn clear(&self) {
+            self.inner.write().blocks.clear();
+        }
+
+        fn chain_head_number_inner(inner: &Inner) -> Option<BlockNumber> {
+            inner.blocks.get(&0).map(|p| p.0.number)
+        }
+
+        pub fn get_block(
+            &self,
+            ancestor: &BlockPtr,
+            offset: BlockNumber,
+        ) -> Option<(BlockPtr, Option<json::Value>)> {
+            Self::get_block_inner(&*self.inner.read(), ancestor, offset)
+        }
+
+        fn get_block_inner(
+            inner: &Inner,
+            ancestor: &BlockPtr,
+            offset: BlockNumber,
+        ) -> Option<(BlockPtr, Option<json::Value>)> {
+            // Check if the ancestor itself is in the cache.
+            let ancestor_offset_from_head = Self::chain_head_number_inner(inner)? - ancestor.number;
+            match inner
+                .blocks
+                .get(&ancestor_offset_from_head)
+                .map(|p| p.0.clone())
+            {
+                None => return None,
+                Some(ptr) if ptr.hash != ancestor.hash => return None,
+                _ => {}
+            }
+
+            let offset_from_head = ancestor_offset_from_head + offset;
+            inner.blocks.get(&offset_from_head).cloned()
+        }
+
+        /// Inserts this block into the cache, if close enough to the chain
+        /// head. If not, it's a no-op.
+        pub fn set_block(&self, ancestor: BlockPtr, ptr: BlockPtr, data: &json::Value) {
+            Self::set_block_inner(&mut *self.inner.write(), ancestor, ptr, data)
+        }
+
+        fn set_block_inner(
+            inner: &mut Inner,
+            ancestor: BlockPtr,
+            ptr: BlockPtr,
+            data: &json::Value,
+        ) {
+            // We must have a chain head to cache blocks.
+            let chain_head_number = if let Some(n) = Self::chain_head_number_inner(inner) {
+                n
+            } else {
+                return;
+            };
+
+            let offset_from_head = chain_head_number - ptr.number;
+            // We only cache blocks that are close to the chain head.
+            if offset_from_head >= inner.capacity as _ {
+                return;
+            }
+            // We only cache blocks if their ancestor is in the cache (chain head excluded).
```

Comment:
The impact of this isn't entirely clear to me; could you add a comment why we do that?

Reply:
The comment wasn't really truthful, in retrospect. I tried to make a better job at addressing the issue in the comment and I'll probably add some others. We only want to add blocks to the cache if a child of that block is already present in the cache, because that's the easiest way to ensure the block is part of the main chain with the correct head. Otherwise, we might end up storing uncle blocks.